### ISSUE TITLE
Cop idea: discourage the use of `assert_`/`refute_` Minitest matchers

### ISSUE THREAD

**Issue by pirj** [rubocop/rubocop-rspec#1485]
[Suggested by @Darhazer](https://github.com/rubocop/rubocop-rspec/pull/1456#issuecomment-1315150349)

RSpec Expectations [allows using](https://github.com/rspec/rspec-expectations/blob/a85a42cb4c4c700b7591e4dc2cb27089bdf425c4/lib/rspec/expectations/minitest_integration.rb) [`assert_`/`refute_` Minitest matchers](http://docs.seattlerb.org/minitest/Minitest/Assertions.html)
However, RSpec comes with its matchers that pretty much cover the same:
```diff
- assert_equal a, b
+ expect(a).to eq(b)
```

A notable difference is that Minitest allows for an optional failure message.
It's a lesser known RSpec feature, [but it allows it, too](https://github.com/rspec/rspec-expectations/blob/6ebb53150c1b1a629e45a4649fa7965e6a95e39a/lib/rspec/expectations/expectation_target.rb#L63):
```diff
- assert_equal a, b, "must be equal"
+ expect(a).to(eq(b), "must be equal")
```

**Comment by bquorning** [comment]
Do we know _why_ rspec-expectations provide this feature? Is it going to be deprecated?

One of the [non-goals of rubocop-rspec](https://docs.rubocop.org/rubocop-rspec/#non-goals-of-rubocop-rspec) is

> Enforcing should vs. expect syntax

and I wonder if the suggested cop doesn’t fall under this category.

**Comment by pirj** [comment]
Fair enough.
RSpeс has a [сonfiguration option for that](https://relishapp.com/rspec/rspec-core/v/3-9/docs/expectation-framework-integration/configure-expectation-framework#configure-minitest-assertions):
```ruby
RSpec.configure do |config|
  config.expect_with :minitest
end
```
which is [exclusively using `rspec-expectations` in the default project `spec/spec_helper.rb` initializer](https://github.com/rspec/rspec-core/blob/6dede8ae1b9cce6e1de4baa912e77c97b587b446/lib/rspec/core/project_initializer/spec/spec_helper.rb#L20) (and most of the projects, I assume):
```ruby
config.expect_with :rspec do |expectations|
```

If with this it's still possible to use Minitest's assertions, there's probably something wrong.

However, `rspec-rails` is using Rails' Minitests assertions internally, and I can't tell off the top of my head if those can be encapsulated and unintentionally leak, or there's just no simple way to encapsulate them and use `rspec-expectations`' matchers exclusively.

**Comment by pirj** [comment]
Indeed, it's [coming from `rspec-rails`](https://github.com/rspec/rspec-rails/blob/2eba2c453ce9ab4eb951925573a142744fb1e0ba/lib/rspec/rails/example/rails_example_group.rb#L13):
```ruby
module RSpec
  module Rails
    # @api public
    # Common rails example functionality.
    module RailsExampleGroup
      # ...
      include RSpec::Rails::MinitestAssertionAdapter
```
I'll give it a shot to only include it if `expect_with` is configured to also allow `:minitest`.

**Comment by pirj** [comment]
https://github.com/rspec/rspec-rails/pull/2639

**Comment by pirj** [comment]
[Fails](https://github.com/rspec/rspec-rails/actions/runs/3503218291/jobs/5868109649), e.g.:
```ruby
1) UploadsController DELETE #destroy redirects to the uploads list
     Failure/Error: expect(response).to redirect_to(uploads_url)

     NoMethodError:
       undefined method `assert' for #<RSpec::ExampleGroups::UploadsController::DELETEDestroy "redirects to the uploads list" (./spec/controllers/uploads_controller_spec.rb:138)>
     # ./spec/controllers/uploads_controller_spec.rb:141:in `block (3 levels) in <top (required)>'
```

If it's impossible to restrict this on RSpec level, I suppose we're left with static analysis.
Unless I misunderstood your argument, @bquorning

**Comment by pirj** [comment]
There are also a few aliases that Rails defines [here](https://github.com/rails/rails/blob/caaac482d5f430b29bb4c80e780fa70e65017755/activesupport/lib/active_support/test_case.rb#L138). I guess the cop could make this list configurable, and `rubocop-rspec-rails` would amend the list supplying those Rails ones? But how would auto-correction work? Or is it better to somehow make the cop extensible from other extensions?

**Comment by bquorning** [comment]
I am still not sure which problem this PR fixes. Minitest matchers are not loaded by default. This example:
```
❯ ruby -r rspec/autorun -e 'describe "foo" do; it "bar" do; assert_equal(1, 2); end; end'
```
gives me the error ```undefined method `assert_equal' for ...```

**Comment by pirj** [comment]
They are if you use `rspec-rails`.

It makes sense to re-target this cop to be `RSpec/Rails/MinitestAssertions` and eventually moving it to `rubocop-rspec-rails` under `RSpecRails/` namespace.

**Comment by pirj** [comment]
Let's keep this open, since there are a few more matchers than can be covered.

**Comment by mvz** [comment]
Please note that the order of the arguments of `assert_equals` is reversed from the `expect` syntax: The actual is the second argument. See http://docs.seattlerb.org/minitest/Minitest/Assertions.html#method-i-assert_equal

**Comment by pirj** [comment]
test-unit assertions, for completeness https://www.rubydoc.info/gems/test-unit/2.3.2/Test/Unit/Assertions

**Comment by G-Rath** [comment]
Also here's the Rails aliased assertion methods: https://github.com/rails/rails/blob/main/activesupport/lib/active_support/test_case.rb#L153
